File: auto_verbs/heavy-anti-bot-sites/tipranks_com/tipranks_search.py
import os
import sys
import shutil
from dataclasses import dataclass, field
from typing import List

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
from cdp_utils import get_free_port, get_temp_profile_dir, launch_chrome, wait_for_cdp_ws
from playwright_debugger import checkpoint


@dataclass(frozen=True)
class TipranksSearchRequest:
    ticker: str = "AAPL"


@dataclass
class TipranksStockItem:
    ticker: str = ""
    company_name: str = ""
    analyst_consensus: str = ""
    price_target: str = ""
    smart_score: str = ""
    num_analysts: str = ""
    sector: str = ""


@dataclass
class TipranksSearchResult:
    stock: TipranksStockItem = field(default_factory=TipranksStockItem)
    error: str = ""


def tipranks_search(page, request: TipranksSearchRequest) -> TipranksSearchResult:
    result = TipranksSearchResult()
    try:
        url = f"https://www.tipranks.com/stocks/{request.ticker}"
        page.goto(url, wait_until="domcontentloaded", timeout=30000)
        page.wait_for_timeout(5000)

        checkpoint(page, "Stock page loaded")

        stock_data = page.evaluate("""() => {
            const getText = (selectors) => {
                for (const sel of selectors) {
                    const el = document.querySelector(sel);
                    if (el && el.textContent.trim()) return el.textContent.trim();
                }
                return '';
            };

            return {
                ticker: getText(['[class*="ticker"], [data-testid="ticker"]']),
                company_name: getText(['h1', '[class*="companyName"], [class*="company-name"]']),
                analyst_consensus: getText(['[class*="consensus"], [class*="rating"]']),
                price_target: getText(['[class*="priceTarget"], [class*="price-target"]']),
                smart_score: getText(['[class*="smartScore"], [class*="smart-score"]']),
                num_analysts: getText(['[class*="numAnalysts"], [class*="analyst-count"]']),
                sector: getText(['[class*="sector"], [class*="industry"]']),
            };
        }""")

        result.stock = TipranksStockItem(**stock_data)

        checkpoint(page, f"Extracted data for {result.stock.ticker}")

    except Exception as e:
        result.error = str(e)
    return result


def test_func():
    port = get_free_port()
    profile_dir = get_temp_profile_dir()
    chrome_proc = launch_chrome(port, profile_dir)
    ws_url = wait_for_cdp_ws(port)

    from playwright.sync_api import sync_playwright
    pw = sync_playwright().start()
    browser = pw.chromium.connect_over_cdp(ws_url)
    context = browser.contexts[0]
    page = context.pages[0] if context.pages else context.new_page()

    try:
        request = TipranksSearchRequest()
        result = tipranks_search(page, request)
        s = result.stock
        print(f"Ticker: {s.ticker}")
        print(f"Company: {s.company_name}")
        print(f"Consensus: {s.analyst_consensus}")
        print(f"Price Target: {s.price_target}")
        print(f"Smart Score: {s.smart_score}")
        print(f"Analysts: {s.num_analysts}")
        print(f"Sector: {s.sector}")
        if result.error:
            print(f"Error: {result.error}")
    finally:
        browser.close()
        pw.stop()
        chrome_proc.terminate()
        shutil.rmtree(profile_dir, ignore_errors=True)


def run_with_debugger():
    test_func()


if __name__ == "__main__":
    run_with_debugger()

File: auto_verbs/heavy-anti-bot-sites/tipranks_com/tipranks_search.js
const { Stagehand } = require("@browserbasehq/stagehand");
const { PlaywrightRecorder, setupLLMClient } = require("../../stagehand-utils");
const fs = require("fs");
const path = require("path");

const CFG = {
  url: "https://www.tipranks.com/stocks/{ticker}",
  ticker: "AAPL",
  outputDir: __dirname,
};

function genPython(cfg, recorder) {
  const pyCode = recorder.generatePythonFromActions(cfg);
  const pyFile = path.join(cfg.outputDir, "tipranks_search.py");
  fs.writeFileSync(pyFile, pyCode, "utf-8");
  console.log(`Python file written to ${pyFile}`);
}

(async () => {
  const stagehand = new Stagehand({
    env: "LOCAL",
    verbose: 1,
    headless: false,
  });
  await stagehand.init();
  const recorder = new PlaywrightRecorder();
  const page = stagehand.context.pages()[0];

  try {
    const stockUrl = CFG.url.replace("{ticker}", encodeURIComponent(CFG.ticker));
    await page.goto(stockUrl, { waitUntil: "domcontentloaded", timeout: 30000 });
    await page.waitForTimeout(5000);

    recorder.addStep("navigate", { url: stockUrl });

    const results = await stagehand.extract(
      `Extract the stock analyst rating information from the TipRanks page for ticker ${CFG.ticker}. Extract the ticker, company_name, analyst_consensus, price_target, smart_score, num_analysts, and sector.`,
      {
        ticker: "string",
        company_name: "string",
        analyst_consensus: "string",
        price_target: "string",
        smart_score: "string",
        num_analysts: "string",
        sector: "string",
      }
    );

    recorder.addStep("extract", {
      description: "Extract stock analyst ratings",
      data: results,
    });

    const outputFile = path.join(CFG.outputDir, "tipranks_search_results.json");
    fs.writeFileSync(outputFile, JSON.stringify(results, null, 2), "utf-8");
    console.log(`Results saved to ${outputFile}`);
    console.log(JSON.stringify(results, null, 2));

    genPython(CFG, recorder);
  } catch (err) {
    console.error("Error:", err.message);
  } finally {
    await stagehand.close();
  }
})();
